### PR TITLE
📝 Add docstrings to `types.py`

### PR DESCRIPTION
**Pull request by Kludex**
Closes DOC-18

Selected Reviewer: @hramezani

### CODE REVIEW

**Comment by Kludex** [comment]
please review

File: pydantic/types.py [hunk 1]
```diff
@@ -1026,39 +1158,86 @@ def __repr__(self) -> str:
...
+        """
         ...
 
 
 class Base64Encoder(EncoderProtocol):
+    """Base64 encoder."""
+
     @classmethod
     def decode(cls, data: bytes) -> bytes:
+        """Decode the data from bytes to a base64 encoded bytes.
```
Comment (lig) [inline pydantic/types.py:1202]: From base64 encoded bytes to original bytes. Encoded state is base64, decoded is raw data.

BTW, the `encode` method has a correct doc string.

**Review by lig — CHANGES_REQUESTED** [review]

File: pydantic/types.py (hunk 1 above)
Comment (lig) [reply pydantic/types.py]:
```suggestion
        """Decode the data from base64 encoded bytes to original bytes data.
```

File: pydantic/types.py [hunk 2]
```diff
@@ -936,6 +1048,18 @@ def condate(
     lt: date | None = None,
     le: date | None = None,
 ) -> type[date]:
+    """A wrapper for date that adds constraints.
+
+    Args:
+        strict: Whether to validate the value in strict mode. Defaults to `None`.
```
Comment (hramezani) [inline pydantic/types.py:1054]: To be consistent with the other docstrings
```suggestion
        strict: Whether to validate the date value in strict mode. Defaults to `None`.
```

File: pydantic/types.py [hunk 3]
```diff
@@ -128,6 +129,19 @@ def conint(
     le: int | None = None,
     multiple_of: int | None = None,
 ) -> type[int]:
+    """A wrapper around `int` that allows for additional constraints.
+
+    Args:
+        strict: Whether to validate the integer in strict mode.
```
Comment (hramezani) [inline pydantic/types.py:135]: Shouldn't we add the default value here?
```suggestion
        strict: Whether to validate the integer in strict mode. Defaults to `None`.
```
Same for other docstrings

File: pydantic/types.py [hunk 4]
```diff
@@ -1026,39 +1158,86 @@ def __repr__(self) -> str:
 
 
 class EncoderProtocol(Protocol):
+    """Protocol for encoding and decoding data to and from bytes."""
+
     @classmethod
     def decode(cls, data: bytes) -> bytes:
-        """Can throw `PydanticCustomError`."""
+        """Decode the data using the encoder.
```
[the remark this answers is not in the document]
Comment (Kludex) [reply pydantic/types.py:1165]: Well, actually, the `EncoderProtocol` defines how to decode and encode. 🤔 

@lig what do you want me to write here?

[the remark this answers is not in the document]
Comment (lig) [reply pydantic/types.py:1165]: I'd just left "Decode the data" and "Encode the data" for the opposite method. This is a protocol definition and it's up to user to decide what to use for the implementation.

**Review by lig — APPROVED** [review]

File: pydantic/types.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kludex) [reply pydantic/types.py:1165]: '-'
```suggestion
        """Decode the data.
```

[the remark this answers is not in the document]
Comment (Kludex) [reply pydantic/types.py:1165]: If we are not willing to change the `EncoderProtocol` name, I think it's better if we maintain the "encoder" here. 🤔

**Pull request merged**